llvm/docs/DirectX/DXILOpTableGenDesign.rst

```diff
@@ -0,0 +1,224 @@
+  def SM6_0 : DXILShaderModel<6, 0>;
+  def SM6_2 : DXILShaderModel<6, 2>;
+
+DXIL Class
+^^^^^^^^^^
+
+Each DXIL Op belongs to a class represented by ``dxil_class`` field value. A
+DXIL class represents DXIL Ops with the same function prototype (or signature).
+This is represented using the following TableGen class.
+
+.. code-block::
+
+  class DXILOpClass<list<LLVMType> OpSig> {
+    list<LLVMType> OpSignature = OpSig;
+  }
+
+Each of the valid classes is represented by a concrete TableGen record.
+For example, following is the definition of the ``unary`` class
+
+.. code-block::
+
+  def unary : DXILOpClass<[llvm_any_ty, LLVMMatch<0>]>;
+
+
+Overload Types
+^^^^^^^^^^^^^^
+
+Each DXIL Op has a set of valid overload types denoted by ``oload_types``.
+Valid overload types of a DXIL OP are represented as a list. However, overload
+types supported by DXIL Ops may vary depending on minimum target shader model
+version. So, the list of supported overload types are tagged with the minimum
+shader model in which they are valid for the DXIL Op being specified.
+Following TableGen class is defined to encapsulate such as representation.
+
+.. code-block::
+
+  class DXILOpOverload<DXILShaderModel minsm, list<LLVMType> overloads> {
+    DXILShaderModel ShaderModel = minsm;
+    list<LLVMType> OpOverloads = overloads;
+  }
+
+Specification of DXIL Operation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Specification of DXIL Op is abstracted using the following TableGen class to
+represent the core attributes outlined in  `DXIL Operation Attributes`_ section.
+
+.. code-block::
+
+  // Abstraction DXIL Operation to LLVM intrinsic
+  class DXILOpMappingBase {
+    int OpCode = 0;                      // Opcode of DXIL Operation
+    DXILOpClass OpClass = UnknownOpClass;// Class of DXIL Operation.
+    Intrinsic LLVMIntrinsic = ?;         // LLVM Intrinsic DXIL Operation maps to
+    list<DXILOpOverload> OpOverloadTypes = ?; // Valid overload type
+                                       // of DXIL Operation
+    string Doc = "";                     // A short description of the operation
+  }
+
+Following is a convenience TableGen class that inherits from ``DXILOpMappingBase``
+with templatized parameters. It is used to define various DXIL Ops.
+
+.. code-block::
+
+  class DXILOpMapping<int opCode,
+                    Intrinsic intrinsic,
+                    list<DXILOpOverload> overloadTypes,
+                    string doc> : DXILOpMappingBase {
+    int OpCode = opCode;
+    Intrinsic LLVMIntrinsic = intrinsic;
+    list<DXILOpOverload> OpOverloadTypes = overloadTypes;
+    string Doc = doc;
+  }
+
+The DXIL Op ``Sin`` is defined as follows:
+
+.. code-block::
+
+  let OpClass = unary in
+    def Sin  : DXILOpMapping<13, int_sin,
+                             [DXILOpOverload<SM6_3, [llvm_half_ty, llvm_float_ty]>,
+                              DXILOpOverload<SM6_0, [llvm_float_ty]>],
+                             "Returns sine(theta) for theta in radians.">;
```

Comment:
There is one more odd availability case that this notation doesn't handle (although I suspect it can be extended to do so). The `ddx` intrinsic is a good example:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-ddx

`ddx` was added for pixel shaders in SM 2.x, but was not added to compute shaders until SM 6.6:

https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_6_Derivatives.html

The overload is the same, but the availability is by-stage and changes in later versions.

Reply:
> There is one more odd availability case that this notation doesn't handle (although I suspect it can be extended to do so). The `ddx` intrinsic is a good example:
...
> `ddx` was added for pixel shaders in SM 2.x, but was not added to compute shaders until SM 6.6:
...
> The overload is the same, but the availability is by-stage and changes in later versions.

Thanks! I plan to extend the specification ability to include Shader Kind as another refinement criterion in a (near-term) future PR.